### DISCUSSION TITLE
Error with Custom Trained YOLOV10 Model: 'dict' object has no attribute 'shape'

### DISCUSSION THREAD

**Discussion by Sukesh-babu** [ultralytics/ultralytics#13586] [Q&A]
Hi everyone,
I'm experiencing an issue with a custom trained YOLOV10 model. When I run inference using a pretrained model, everything works fine. However, when I use my custom trained model, I get the following error: 
AttributeError: 'dict' object has no attribute 'shape'

I trained a YOLOv10 model using custom data. The code i used for training is as follows

from ultralytics import YOLO
import torch
torch.cuda.empty_cache()
print("CUDA is available : ", torch.cuda.is_available())
def main():
    model = YOLO('weights/yolov10n.pt')
    results = model.train(data="custom_config.yaml", epochs=1, batch=16, save=True, save_period=1, workers=8, 	resume=False, 
     plots=True, device=0) # batch limit for HP Z4 is '32' ('64' returns 	torch.cuda.OutofMemoryError)
   
if __name__ == '__main__':
    main()

The training was success full.
During inference, I encounter a An error occurred during prediction: 'dict' object has no attribute 'shape'.
#### Steps to Reproduce
1. Train the YOLOv10 model with custom data.
2. Run the inference script as follows:

```python
import os
import cv2
from ultralytics import YOLO

model = YOLO(f"custom_best.pt")
model.predict(source="D:/YOLOV10_TR/images/", save=True, classes=[0])

## I used the foolowing scripts also for inference

```python
import os
import torch
from ultralytics import YOLO

def main():
    model_path = "custom_best.pt"
    model = YOLO(model_path)
    source_path = "D:/YOLOV10_TR/images/"

    try:
        results = model.predict(source=source_path, save=True, classes=[0])
        predictions = results['predictions']  # Error occurs here
        predictions = torch.tensor(predictions)
        nms_results = non_max_suppression(predictions)
        print("NMS Results:", nms_results)
    except Exception as e:
        print("An error occurred during prediction:", e)

if __name__ == '__main__':
    main()
While runnig the above scripts(both), the following error
An error occurred during prediction: 'dict' object has no attribute 'shape'.

Environment:
 Python version: 3.9.19
 YOLO version: V10n
Additional Information:
I have ensured that the model is correctly loaded and the paths are correct. The same script works perfectly with the pretrained model.

Any help or pointers to resolve this issue would be greatly appreciated. Thank you!

**Comment by pderrenger** [comment]
@Sukesh-babu hi there,

Thank you for reaching out and providing detailed information about the issue you're encountering with your custom trained YOLOv10 model. Let's work through this together.

### Steps to Reproduce and Debug

1. **Reproducible Code Example**:
   Your provided code is quite comprehensive, but to ensure we can reproduce the issue accurately, could you please confirm if the error persists with the latest versions of `torch` and `ultralytics`? If not, please upgrade your packages and try again:
   ```bash
   pip install --upgrade torch ultralytics
   ```

2. **Inference Script**:
   The error `AttributeError: 'dict' object has no attribute 'shape'` suggests that the `results` object is being treated as a dictionary, which is not expected. The `predict` method should return a list of `Results` objects. Let's modify your inference script to handle this correctly:
   ```python
   import os
   from ultralytics import YOLO

   def main():
       model_path = "custom_best.pt"
       model = YOLO(model_path)
       source_path = "D:/YOLOV10_TR/images/"

       try:
           results = model.predict(source=source_path, save=True, classes=[0])
           for result in results:
               print(result)  # Display each result
               # Access the boxes, masks, keypoints, etc. from the result object
               boxes = result.boxes
               print("Boxes:", boxes)
       except Exception as e:
           print("An error occurred during prediction:", e)

   if __name__ == '__main__':
       main()
   ```

3. **Ensure Correct Usage**:
   Ensure that the `results` object is being used correctly. The `predict` method returns a list of `Results` objects, not a dictionary. Therefore, accessing `results['predictions']` will not work. Instead, iterate over the `results` list as shown above.

4. **Environment Check**:
   Verify that your environment is correctly set up and that there are no conflicting versions of libraries. You can do this by creating a new virtual environment and installing the necessary packages:
   ```bash
   python -m venv yolov10_env
   source yolov10_env/bin/activate  # On Windows use `yolov10_env\Scripts\activate`
   pip install torch ultralytics
   ```

### Additional Resources
For further details on using YOLOv3 and its variants, you can refer to the [YOLOv3 documentation](https://docs.ultralytics.com/models/yolov3/).

If the issue persists after trying the above steps, please provide any additional error messages or logs that could help us diagnose the problem more effectively.

Thank you for your patience and cooperation. Let's get this resolved! 🚀

**Comment by Sukesh-babu** [comment]
Respected sir;
    I created a new environment and  I ensured that I am using the latest
versions of torch and ultralytics. After that I run the modified version of
the inference script. But the same error: "An error occurred during
prediction: 'dict' object has no attribute 'shape'". The script perfectly
works for pretrained model.

On Fri, Jun 14, 2024 at 10:30 PM Paula Derrenger ***@***.***>
wrote:

> @Sukesh-babu <https://github.com/Sukesh-babu> hi there,
>
> Thank you for reaching out and providing detailed information about the
> issue you're encountering with your custom trained YOLOv10 model. Let's
> work through this together.
> Steps to Reproduce and Debug
>
>    1.
>
>    *Reproducible Code Example*:
>    Your provided code is quite comprehensive, but to ensure we can
>    reproduce the issue accurately, could you please confirm if the error
>    persists with the latest versions of torch and ultralytics? If not,
>    please upgrade your packages and try again:
>
>    pip install --upgrade torch ultralytics
>
>    2.
>
>    *Inference Script*:
>    The error AttributeError: 'dict' object has no attribute 'shape'
>    suggests that the results object is being treated as a dictionary,
>    which is not expected. The predict method should return a list of
>    Results objects. Let's modify your inference script to handle this
>    correctly:
>
>    import osfrom ultralytics import YOLO
>    def main():
>        model_path = "custom_best.pt"
>        model = YOLO(model_path)
>        source_path = "D:/YOLOV10_TR/images/"
>
>        try:
>            results = model.predict(source=source_path, save=True, classes=[0])
>            for result in results:
>                print(result)  # Display each result
>                # Access the boxes, masks, keypoints, etc. from the result object
>                boxes = result.boxes
>                print("Boxes:", boxes)
>        except Exception as e:
>            print("An error occurred during prediction:", e)
>    if __name__ == '__main__':
>        main()
>
>    3.
>
>    *Ensure Correct Usage*:
>    Ensure that the results object is being used correctly. The predict
>    method returns a list of Results objects, not a dictionary. Therefore,
>    accessing results['predictions'] will not work. Instead, iterate over
>    the results list as shown above.
>    4.
>
>    *Environment Check*:
>    Verify that your environment is correctly set up and that there are no
>    conflicting versions of libraries. You can do this by creating a new
>    virtual environment and installing the necessary packages:
>
>    python -m venv yolov10_envsource yolov10_env/bin/activate  # On Windows use `yolov10_env\Scripts\activate`
>    pip install torch ultralytics
>
>
> Additional Resources
>
> For further details on using YOLOv3 and its variants, you can refer to the YOLOv3
> documentation <https://docs.ultralytics.com/models/yolov3/>.
>
> If the issue persists after trying the above steps, please provide any
> additional error messages or logs that could help us diagnose the problem
> more effectively.
>
> Thank you for your patience and cooperation. Let's get this resolved! 🚀
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/ultralytics/discussions/13586#discussioncomment-9776968>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3WQSKMDGCUWJCPVIMKQO5LZHMORXAVCNFSM6AAAAABJKCAYBOVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TONZWHE3DQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
𝐁𝐲
𝐒𝐮𝐤𝐞𝐬𝐡 𝐁𝐚𝐛𝐮 𝐕 𝐒
𝐑𝐞𝐬𝐞𝐚𝐫𝐜𝐡 𝐒𝐜𝐡𝐨𝐥𝐚𝐫
𝐃𝐞𝐩𝐚𝐫𝐭𝐦𝐞𝐧𝐭 𝐨𝐟 𝐂𝐨𝐦𝐩𝐮𝐭𝐞𝐫 𝐒𝐜𝐢𝐞𝐧𝐜𝐞 𝐚𝐧𝐝
𝐄𝐧𝐠𝐢𝐧𝐞𝐞𝐫𝐢𝐧𝐠,
𝐈𝐧𝐝𝐢𝐚𝐧 𝐈𝐧𝐬𝐭𝐢𝐭𝐮𝐭𝐞 𝐨𝐟 𝐈𝐧𝐟𝐨𝐫𝐦𝐚𝐭𝐢𝐨𝐧
𝐓𝐞𝐜𝐡𝐧𝐨𝐥𝐨𝐠𝐲 𝐃𝐞𝐬𝐢𝐠𝐧 𝐀𝐧𝐝 𝐌𝐚𝐧𝐮𝐟𝐚𝐜𝐭𝐮𝐫𝐢𝐧𝐠
(𝐈𝐈𝐈𝐓𝐃𝐌 𝐊𝐚𝐧𝐜𝐡𝐞𝐞𝐩𝐮𝐫𝐚𝐦)
𝐌𝐨𝐛: 𝟗𝟖𝟒𝟕𝟔𝟑𝟎𝟖𝟎𝟔

**Reply by glenn-jocher** [reply]
@Sukesh-babu hi Sukesh,

Thank you for your detailed follow-up and for ensuring you're using the latest versions of `torch` and `ultralytics`. Let's address the issue you're encountering with your custom trained YOLOv10 model.

### Issue Analysis
The error `AttributeError: 'dict' object has no attribute 'shape'` suggests that the `results` object is being treated as a dictionary, which is not expected. The `predict` method should return a list of `Results` objects, not a dictionary.

### Suggested Solution
Let's modify your inference script to handle the `results` object correctly. Here’s an updated version of your script:

```python
import os
from ultralytics import YOLO

def main():
    model_path = "custom_best.pt"
    model = YOLO(model_path)
    source_path = "D:/YOLOV10_TR/images/"

    try:
        results = model.predict(source=source_path, save=True, classes=[0])
        for result in results:
            print(result)  # Display each result
            # Access the boxes, masks, keypoints, etc. from the result object
            boxes = result.boxes
            print("Boxes:", boxes)
    except Exception as e:
        print("An error occurred during prediction:", e)

if __name__ == '__main__':
    main()
```

### Key Points:
1. **Iterate Over Results**: The `predict` method returns a list of `Results` objects. Iterate over this list to access individual results.
2. **Access Attributes**: Use the attributes of the `Results` object (like `boxes`, `masks`, etc.) to get the desired information.

### Environment Check
To ensure there are no conflicting versions of libraries, please create a new virtual environment and install the necessary packages:

```bash
python -m venv yolov10_env
source yolov10_env/bin/activate  # On Windows use `yolov10_env\Scripts\activate`
pip install torch ultralytics
```

### Additional Resources
For further details on using YOLOv10 and its variants, you can refer to the [YOLOv10 documentation](https://docs.ultralytics.com/models/yolov10/).

If the issue persists after trying the above steps, please provide any additional error messages or logs that could help us diagnose the problem more effectively.

Thank you for your patience and cooperation. Let's get this resolved! 🚀

**Comment by Sukesh-babu** [comment]
Respected sir;
                               I checked the environment, and also created
a new environment and runs the modified script. But the same error: An
error occurred during prediction: 'dict' object has no attribute 'shape

On Sat, Jun 15, 2024 at 1:53 PM Glenn Jocher ***@***.***>
wrote:

> @Sukesh-babu <https://github.com/Sukesh-babu> hi Sukesh,
>
> Thank you for your detailed follow-up and for ensuring you're using the
> latest versions of torch and ultralytics. Let's address the issue you're
> encountering with your custom trained YOLOv10 model.
> Issue Analysis
>
> The error AttributeError: 'dict' object has no attribute 'shape' suggests
> that the results object is being treated as a dictionary, which is not
> expected. The predict method should return a list of Results objects, not
> a dictionary.
> Suggested Solution
>
> Let's modify your inference script to handle the results object
> correctly. Here’s an updated version of your script:
>
> import osfrom ultralytics import YOLO
> def main():
>     model_path = "custom_best.pt"
>     model = YOLO(model_path)
>     source_path = "D:/YOLOV10_TR/images/"
>
>     try:
>         results = model.predict(source=source_path, save=True, classes=[0])
>         for result in results:
>             print(result)  # Display each result
>             # Access the boxes, masks, keypoints, etc. from the result object
>             boxes = result.boxes
>             print("Boxes:", boxes)
>     except Exception as e:
>         print("An error occurred during prediction:", e)
> if __name__ == '__main__':
>     main()
>
> Key Points:
>
>    1. *Iterate Over Results*: The predict method returns a list of Results
>    objects. Iterate over this list to access individual results.
>    2. *Access Attributes*: Use the attributes of the Results object (like
>    boxes, masks, etc.) to get the desired information.
>
> Environment Check
>
> To ensure there are no conflicting versions of libraries, please create a
> new virtual environment and install the necessary packages:
>
> python -m venv yolov10_envsource yolov10_env/bin/activate  # On Windows use `yolov10_env\Scripts\activate`
> pip install torch ultralytics
>
> Additional Resources
>
> For further details on using YOLOv10 and its variants, you can refer to
> the YOLOv10 documentation <https://docs.ultralytics.com/models/yolov10/>.
>
> If the issue persists after trying the above steps, please provide any
> additional error messages or logs that could help us diagnose the problem
> more effectively.
>
> Thank you for your patience and cooperation. Let's get this resolved! 🚀
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/ultralytics/discussions/13586#discussioncomment-9780604>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3WQSKIC7ONSVXHD3TK4QETZHP2WPAVCNFSM6AAAAABJKCAYBOVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TOOBQGYYDI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
𝐁𝐲
𝐒𝐮𝐤𝐞𝐬𝐡 𝐁𝐚𝐛𝐮 𝐕 𝐒
𝐑𝐞𝐬𝐞𝐚𝐫𝐜𝐡 𝐒𝐜𝐡𝐨𝐥𝐚𝐫
𝐃𝐞𝐩𝐚𝐫𝐭𝐦𝐞𝐧𝐭 𝐨𝐟 𝐂𝐨𝐦𝐩𝐮𝐭𝐞𝐫 𝐒𝐜𝐢𝐞𝐧𝐜𝐞 𝐚𝐧𝐝
𝐄𝐧𝐠𝐢𝐧𝐞𝐞𝐫𝐢𝐧𝐠,
𝐈𝐧𝐝𝐢𝐚𝐧 𝐈𝐧𝐬𝐭𝐢𝐭𝐮𝐭𝐞 𝐨𝐟 𝐈𝐧𝐟𝐨𝐫𝐦𝐚𝐭𝐢𝐨𝐧
𝐓𝐞𝐜𝐡𝐧𝐨𝐥𝐨𝐠𝐲 𝐃𝐞𝐬𝐢𝐠𝐧 𝐀𝐧𝐝 𝐌𝐚𝐧𝐮𝐟𝐚𝐜𝐭𝐮𝐫𝐢𝐧𝐠
(𝐈𝐈𝐈𝐓𝐃𝐌 𝐊𝐚𝐧𝐜𝐡𝐞𝐞𝐩𝐮𝐫𝐚𝐦)
𝐌𝐨𝐛: 𝟗𝟖𝟒𝟕𝟔𝟑𝟎𝟖𝟎𝟔

**Reply by glenn-jocher** [reply]
@Sukesh-babu hi Sukesh,

Thank you for your detailed follow-up and for ensuring you're using the latest versions of `torch` and `ultralytics`. Let's address the issue you're encountering with your custom-trained YOLOv10 model.

### Issue Analysis
The error `AttributeError: 'dict' object has no attribute 'shape'` suggests that the results object is being treated as a dictionary, which is not expected. The `predict` method should return a list of `Results` objects, not a dictionary.

### Suggested Solution
Let's modify your inference script to handle the results object correctly. Here’s an updated version of your script:

```python
import os
from ultralytics import YOLO

def main():
    model_path = "custom_best.pt"
    model = YOLO(model_path)
    source_path = "D:/YOLOV10_TR/images/"

    try:
        results = model.predict(source=source_path, save=True, classes=[0])
        for result in results:
            print(result)  # Display each result
            # Access the boxes, masks, keypoints, etc. from the result object
            boxes = result.boxes
            print("Boxes:", boxes)
    except Exception as e:
        print("An error occurred during prediction:", e)

if __name__ == '__main__':
    main()
```

### Key Points:
1. **Iterate Over Results**: The `predict` method returns a list of `Results` objects. Iterate over this list to access individual results.
2. **Access Attributes**: Use the attributes of the `Results` object (like `boxes`, `masks`, etc.) to get the desired information.

### Environment Check
To ensure there are no conflicting versions of libraries, please create a new virtual environment and install the necessary packages:

```bash
python -m venv yolov10_env
source yolov10_env/bin/activate  # On Windows use `yolov10_env\Scripts\activate`
pip install torch ultralytics
```

### Additional Resources
For further details on using YOLOv10 and its variants, you can refer to the [YOLOv10 documentation](https://docs.ultralytics.com/models/yolov10/).

If the issue persists after trying the above steps, please provide any additional error messages or logs that could help us diagnose the problem more effectively.

Thank you for your patience and cooperation. Let's get this resolved! 🚀
